### PR TITLE
Feature: ui grid filtering

### LINKED ISSUE #336: (Design) Patients Registry
This issue describes the patient registry module proposed/ described in https://github.com/IMA-WorldHealth/bhima-2.X/issues/278. 

The patients registry (records) will be the central place in BHIMA to search for a patient, it will be a tool that can be used to check in a returning patient or looking up any patient that has been registered in the past. 

These proposed designs show a complex final state - the core of this module should load patient records and allow filtering on columns. 

Options to search for patients:
- When was the patient registered? - from date, to date 
- When was the patients last bill? - from date, to date
- How old is the patient? - from age, to age
- Gender of the patient? 
- Origin location? 
- Current location? 
- Searches on names

If there are any additional ideas on parameters to search for patients they can be commented and added here.

**Patient Registry Core**
patient registry

I propose that this module uses a UI Grid that allows sorting/ filtering on the current patient records search. On the initial page load the grid will show the most recently registered or visiting patients. 

**Patient Registry Search**
patient registry search

A modal search service will be used to refine the results - this will make a new request to the server for patient records based on the search parameters. It should also provide basic form validation.

**Patient Registry Search View**
patient registry filtered

The current search parameters should be shown to the user somehow. This mockup shows a filter bar that could be used across a number of modules like this. How filters/ searches are shown should be discussed.

### LINKED ISSUE #492: (Proposal) UI Grid Filters
The current filtering of UI Grids is bulky and difficult to work with.  Some problems with it:
1. Cannot selectively add or remove filters.  The modal applies all filters at once, and removes them all when filtering again.
2. Filters should ideally be stored as JSON objects to be sent as query params to the server.  They are currently stored in an array that is difficult to work with on the client.

**Proposal**
I propose that we model filters as a JSON object with `column:value` pairs.  Like this:

``` js
// filters to be applied to the grid.
let filters = {
  name: "Christopher",
  limit: "10",
  sex: "M"
};
```

I also propose that we implement add/remove functionality that can remove or add a single filter at a time.  The UI would look like this:
ui grid filters no filters
_Fig 1: No filters applied._

ui grid filters filters applied
_Fig 2: Filtering with a single filter + limit._

ui grid filters multiple filters
_Fig 3: Filtering with multiple, additive filters._

When the user clicks the "X" icon on a filter, it is removed, and the data is refreshed from the database.

### PR DESCRIPTION
**Pull request by jniles**
For full details of this change, see the commit messages.  The changes are best explained via screenshots:

**Screenshots**
thepatientregistry
_Fig 1: The Unfiltered Patient Registry_

filterform
_Fig 2: The New Filter Form_

filteredregistry
_Fig 3: The New Filter Bar_

filteredpdfreport
_Fig 4: Filters appear on rendered PDF report_

**Major Features**
1. Clicking "x" on any filter will remove it and re-query the database.
2. The filters are cached, so that page refresh remembers the filters.
3. The filter form will be re-populated with filters from the grid when it is opened a second time.
4. The "Clear all Filters" button removes all filters from the grid.
5. The dates on the form have a clearer UI (start, end).
6. The PDF now shows what filters rendered it in a bar identical to the client's bar.

Closes #492.  Closes #336.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by jniles** [comment]
@sfount, could I get a review?

File: client/test/e2e/patient/registry.search.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  });
+
+  // demonstrates that filtering works
+  it(`should find one patient with name "${parameters.name}"`, () => {
+    FU.buttons.search();
+    FU.input('ModalCtrl.params.name', parameters.name);
+    FU.buttons.submit();
+
+    expectNumberOfGridRows(1);
+    expectNumberOfFilters(1);
```
Comment (sfount) [inline client/test/e2e/patient/registry.search.js:61]: :+1:

File: client/src/js/components/bhFiltersApplied.js [hunk 2]
```diff
@@ -1,7 +1,29 @@
...
   bindings: {
-    filters: '='
+    filters: '<',
+    onRemoveFilter: '&'  // fires to remove the filter
   }
 });
+
+bhFiltersAppliedController.$inject = [ '$scope', '$filter' ];
+
+function bhFiltersAppliedController($scope, $filter) {
```
Comment (sfount) [inline client/src/js/components/bhFiltersApplied.js:42]: Could you leave a quick comment explaining what this directive does? 

Just to make sure I have it straight 
- Displays all filters passed in through the filters attribute
- Runs any Angular filter through the $filter service before displaying the value

A very short reasoning behind the $watch may be useful for the next developer working on filters.

**Comment by sfount** [comment]
@jniles This LGTM - very slick and well thought through presentation on the filter bar. 

I've left one note about comments on the directive, if this will go through further iteration soon just let me know - it is not at all blocking on this pull request.

File: client/src/js/components/bhFiltersApplied.js (hunk 2 above)
Comment (jniles) [inline client/src/js/components/bhFiltersApplied.js:42]: Sure, I'll do that in a second and ping you back when it is complete.

**Comment by jniles** [comment]
@sfount, I've made the suggested changes.

**Pull request merged**